magda-gateway/src/oauth2/aaf.ts

```diff
@@ -0,0 +1,133 @@
+import * as express from "express";
+const CustomStrategy = require("passport-custom").Strategy;
+import { Authenticator, Profile } from "passport";
+var jwt = require("jwt-simple");
+import ApiClient from "@magda/typescript-common/dist/authorization-api/ApiClient";
+import createOrGetUserToken from "../createOrGetUserToken";
+import { redirectOnSuccess, redirectOnError } from "./redirect";
+
+export interface aafOptions {
+    authorizationApi: ApiClient;
+    passport: Authenticator;
+    aafClientUri: string;
+    aafClientSecret: string;
+    externalUrl: string;
+}
+
+export default function aaf(options: aafOptions) {
+    const authorizationApi = options.authorizationApi;
+    const passport = options.passport;
+    const aafClientUri = options.aafClientUri;
+
+    /**
+     * Many base64 ultiliy will add a trailing \n to encoded content
+     * Just in case the secret has a trailing \n
+     * Using trim to remove it.
+     * Make sure aafClientSecret is a string to avoid `no toString method`
+     * error from jwt.decode
+     */
+    const aafClientSecret = (options.aafClientSecret
+        ? "" + options.aafClientSecret
+        : ""
+    ).trim();
+
+    const aafSuccessRedirect = `${
+        options.externalUrl
+    }/sign-in-redirect?redirectTo=/`;
+    const aafFailRedirect = `${
+        options.externalUrl
+    }/sign-in-redirect?redirectTo=/signin`;
+
+    if (!aafClientUri) {
+        return undefined;
+    }
+
+    if (!aafClientSecret) {
+        console.log("aafClientSecret is empty!");
+    }
+
+    passport.use(
+        new CustomStrategy(function(req: any, done: any) {
+            var verified_jwt = jwt.decode(
```

Comment:
Can we change these to `const`s or `let`s?

Reply:
Updated~ They are all `const` now~